pkg/xds/server/v3/reconcile.go

```diff
@@ -40,30 +40,71 @@ func (r *reconciler) Reconcile(ctx xds_context.Context, proxy *model.Proxy) erro
 		reconcileLog.Error(err, "failed to generate a snapshot", "node", node, "proxy", proxy)
 		return err
 	}
-	if err := snapshot.Consistent(); err != nil {
-		reconcileLog.Error(err, "inconsistent snapshot", "snapshot", snapshot, "proxy", proxy)
-	}
-	// to avoid assigning a new version every time,
-	// compare with the previous snapshot and reuse its version whenever possible,
+
+	// To avoid assigning a new version every time, compare with
+	// the previous snapshot and reuse its version whenever possible,
 	// fallback to UUID otherwise
 	previous, err := r.cacher.Get(node)
 	if err != nil {
 		previous = envoy_cache.Snapshot{}
 	}
-	snapshot = r.autoVersion(previous, snapshot)
+
+	snapshot, changed := autoVersion(previous, snapshot)
+
+	// Validate the resources we reconciled before sending them
+	// to Envoy. This ensures that we have as much in-band error
+	// information as possible, which is especially useful for tests
+	// that don't actually program an Envoy instance.
+	if changed {
+		for _, resources := range snapshot.Resources {
+			for name, resource := range resources.Items {
+				if err := validateResource(resource.Resource); err != nil {
+					reconcileLog.Error(err, "invalid resource", "name", name, "proxy", proxy)
+					return err
+				}
+			}
+		}
+
+		if err := snapshot.Consistent(); err != nil {
+			reconcileLog.Error(err, "inconsistent snapshot", "snapshot", snapshot, "proxy", proxy)
+			return err
```

Comment:
I think we should just return the error. Not log + error. I know there is the same thing already in line 40, but logging should be only on the top level (`log.Error(err, "OnTick() failed")`), otherwise we are duplicating errors.

Reply:
I removed all the logs except this one. I think that we should keep this because it's the only place that dumps the snapshot itself, which is needed to debug why it is inconsistent (the errors out of go-control-plane are not very helpful for this).